tests/Antlers/Runtime/VariablePriorityTest.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Tests\Antlers\Runtime;
+
+use Statamic\Exceptions\CollectionNotFoundException;
+use Statamic\Exceptions\TaxonomyNotFoundException;
+use Statamic\Modifiers\ModifierNotFoundException;
+use Tests\Antlers\ParserTestCase;
+
+class VariablePriorityTest extends ParserTestCase
+{
+    protected $data = [];
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->data = [
+            'collection' => [
+                'articles' => [
+                    ['title' => 'Test One'],
+                    ['title' => 'Test Two'],
+                ],
+                'handle' => 'The Collection',
+            ],
+            'form' => [
+                'handle' => 'The Form',
+            ],
+            'link' => 'Hello, there!',
+            'taxonomy' => 'The Taxonomy',
+        ];
+    }
+
+    private function render($template)
+    {
+        return $this->renderString($template, $this->data, true);
+    }
+
+    public function test_arrays_take_priority_over_tags()
+    {
+        $template = <<<'EOT'
+{{ collection:articles }}<{{ title }}>{{ /collection:articles }}
+EOT;
+
+        $this->assertSame('<Test One><Test Two>', $this->render($template));
+    }
+
+    public function test_collection_tag_is_still_invoked()
+    {
+        $template = <<<'EOT'
+{{ collection:news }}<{{ title }}>{{ /collection:news }}
+EOT;
+        $this->expectException(CollectionNotFoundException::class);
+        $this->expectExceptionMessage('Collection [news] not found');
+        $this->render($template);
+    }
+
+    public function test_strings_can_be_used_inside_tags_with_similar_names()
+    {
+        $template = <<<'EOT'
+{{ collection from="{collection}" }}{{ /collection }}
+EOT;
+
+        $this->expectException(CollectionNotFoundException::class);
+        $this->expectExceptionMessage('Collection [news] not found');
+        $this->renderString($template, ['collection' => 'news'], true);
+    }
```

Comment:
This feels so magical. (But expected) How does it know you aren't intending to call a `from` modifier on a variable named `collection`?

Reply:
The NodeProcessor now utilizes the following simplified steps to determine if it should process something as a tag or not. If a template author has not forced the item to be a variable using the `$` syntax, or forced it to be a tag by using the `%` syntax, the processor will perform the following checks. In the following, the node's default behavior is determined ahead of time by checking if the node's name matches a registered tag (and is exposed via the `$node->isTagNode` property).

1. Is the variable part of a tag pair? If so, and the resolve value is loopable, we proceed as a variable. If the resolved value is not loopable, we proceed based on the node's default behavior.
2. If the variable is not a tag pair, does the variable existing within the current context? If so, we proceed as a variable. If not, we proceed with the node's default behavior.

There are some additional checks surrounding query builder instances to ensure that they are sent to the `query` tag (unless they appear as part of something like `{{ query_builder.0.title }}` - in these scenarios, the value is simply resolved and the rest of the variable path is processed like any other array).